### ISSUE TITLE
Use Kubernetes API instead of Docker socket

### ISSUE THREAD

**Issue by phillebaba** [microsoft/azure-pipelines-agent#3138]
Mounting a Docker socket to another container is not an optimal solution, as it introduces its own set of security issues. Additionally it is dependent on Docker, which will become more of an issue as Docker is removed from AKS.

It would be interesting if we could implement a KubernetesOperationProvider in the same way that the ContainerOperationProvider exists today. In theory this would not require any changes to the yaml schema as it would be a setting in the agent whether it uses the Kubernetes API to create containers in Pods or if it uses Docker to do the same thing.
https://github.com/microsoft/azure-pipelines-agent/blob/af14232b3754f9962d5e1ffefb9ac8a4a0abd89e/src/Agent.Worker/ContainerOperationProvider.cs 

This solution could also introduce an easier way of doing autoscaling as a single agent could now also in theory spawn multiple jobs at the same time.

Is this a possible feature that could be implemented?

**Comment by kgorskowski** [comment]
Something like this would be quite awesome. The last week I tried to build something like that because we wanted to haven our pipelines more scaleable but at the moment I would say the scale sets are the first class citizens here. One problem I was facing was the "1:1 relation" between agents and jobs. To have an awareness "inside" the cluster if there are jobs in the queue its necessary to "poll" the queue actively or have a pipeline task to start an "on demand" agent task. Looked into running them as jobs and had it kinda working but the whole Docker in Docker, exit graceful, enable cluster upscaling but prevent running job termination was not trivial. In a nutshell: to have a kind of orchestrator spawning one-off agent pods/jobs is exactly was I was hoping to archive :)

**Comment by mauricio-obenaus** [comment]
That would be nice... Jenkins does this and is awesome, works flawlessly.

I did that with some level of success with azure, but now really happy with it, I'm using it for test only.

What I did:
- first passing the --once parameter to the agent, so it exits and the Kubernetes scheduler restart it, so, one agent, one job.
- used an Ubuntu docker in docker, works ok, but need checking the MTU parameter for docker networks.
- with StatefulSet its grate, the agents use the same name with a number suffix.
- to scale the agent I did a small script that pool the azure API to calculate the demand and scale the StatefulSet correctly, to scale down, the script verifies who is the last agent, and scale down just to that agent, preventing that the running agent don't get killed.

What I liked, is that the agent starts in about 20 seconds, with a clean context.

What I didn't like, if you use containers, it always pulls them, an empty agent don't have any images on the cache, that could be resolved using the host docker.  Still to be tested.

And I now this can stop work any time...

**Comment by anhdle14** [comment]
@kgorskowski You can actually do somethings with [pool.demands](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/demands?view=azure-devops&tabs=yaml#manually-entered-demands).

This solution works great with K8S and Docker. But the same thing can be achieved by a queue, serverless, and VMs. I only implemented the prior not the later.

There are 2 types of agent (both are azure-pipeline agents)

- Orchestrator (Init - Always on)
- Worker (Build - On-demand)

And the sample azure-pipelines.yaml:

```yaml
stages:
- stage: PrepareBuildAgent # 1
  pool:
    name: agentPoolName
    demand: AgentType -eq Init
  jobs:
  - job:
    steps:
    - step: helm install worker-agent
- stage: BuildorRelease # 1..n
  pool:
    name: agentPoolName
    demand: AgentType -eq Worker-nodejs/dotnet/python
- stage: CleanUp # n
  pool:
    name: agentPoolName
    demand: AgentType -eq Init
  jobs:
  - job:
    steps:
    - step: helm uninstall worker-agent
```

---
Basically #1 should create the worker agents somehow. I used Helm and pass the correct docker images and replicas to create build agents.

And step #n should always run to clean up resources that were being created in #1. I can share more if you are interested. 

One downside for AKS is docker-in-docker is not supported. So you have to use `az acr build` and caching docker images somehow. I don't think anyone should go down this hole, you can just use jenkins or tekton.cd. Unless there is an enterprise that stuck with Azure DevOps.

**Comment by sennerholm** [comment]
@anhdle14 It would be interesting to know more. I currently run something that is based on: https://github.com/anoff/autoscaling-ci-agents-on-aks
but that seems dead, and all other projects I have found to run some azure agents on k8s also. Something was on the azure devops roadmap before, but I don't know the status of it.

**Comment by anhdle14** [comment]
@sennerholm Yes this is basically the gist of my solution as well. The best way to go forward is to write a complete K8S API controller and analyze the current number of jobs from either Azure API or manually. However, I think it is just better to use [Tekton.cd](https://tekton.dev) at this point. 

My solution from above is simpler but easier to execute. I don't think I am the only one thing of it, I got the similar ideas from StackOverFlow but couldn't recall the source. But I will share everything here:

# Setting up the agents

1. Base agent (nothing fancy here, just get the sample from [Azure Docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#create-and-build-the-dockerfile-1) on self-hosted agents with Docker)
2. Init agent
    - Build on top of base agent
    - Adding Helm or Kubectl CLI in order to call K8S API to create/delete build agents
        - i.e. `Helm upgrade --install <name> --namespace <agent pool name> --set agent.type <agent type (nodejs> --replicas <number of worker agents> /charts/worker-agent`
        - i.e. `Helm uninstall <name> --namespace <agent pool name>`
    - This agent should always on. 
3. Build/Worker agents
    - Build on top of base agent
    - Adding any nodejs, python, java SDK tools

> You can tag the worker agents with build-agent:nodejs, build-agent:python so it is easier to have specific worker agent. 

# Setting up Azure DevOps

- We have multiple projects so I think it is best that you have namespace separation for each project. So it is easy to clean up any agent pool. 
- Like I mentioned you can basically use [pool.demands](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/demands?view=azure-devops&tabs=yaml#manually-entered-demands)to run a job in either Init agent or Worker agent.
- Not automated as https://github.com/anoff/autoscaling-ci-agents-on-aks, instead it is a manual process.

# Setting up the Azure-Pipelines.yaml 

```yaml
stages:
- stage: PrepareBuildAgent # 1
  pool:
    name: agentPoolName
    demand: AgentType -eq init
  jobs:
  - job:
    steps:
    - step: helm install worker-agent --set agent.type=nodejs
- stage: BuildorRelease # 1..n
  pool:
    name: agentPoolName
    demand: AgentType -eq nodejs
- stage: CleanUp # n
  pool:
    name: agentPoolName
    demand: AgentType -eq init
  jobs:
  - job:
    steps:
    - step: helm uninstall worker-agent
```

**Comment by sennerholm** [comment]
@anhdle14 Thanks for the quick reply and your gist.

> @sennerholm Yes this is basically the gist of my solution as well. The best way to go forward is to write a complete K8S API controller and analyze the current number of jobs from either Azure API or manually. However, I think it is just better to use [Tekton.cd](https://tekton.dev) at this point.

Currently the policy is azure devops, so I can't go the Tekton.Cd path. 
The Anoff code is doing some analyzing of the number of jobs etc, but it's not maintained and needing some improvement to fit our needs. I
Anyone who knows about any living project for doing an analyzing of the Azure API and integrate it to K8s? I prefer to try to improve something already existing instead of reinventing the wheel.

**Comment by anhdle14** [comment]
Not stale, AKS needs to support some alternative to mounting docker-socket. Please check tekton or argo-workflow project.

**Comment by SebastianSchuetze** [comment]
I wrote a blog post about this. Works with KEDA and Azure Container Apps natively and scalable

https://www.razorspoint.com/2021/11/19/scalable-container-based-azure-pipelines-pools-with-azure-container-apps/

Maybe that solves the problem?

**Comment by anatolybolshakov** [comment]
Hi everyone! Thanks for reporting, we are currently working on more prioritized issues, but will get back to this one once be able to.

**Comment by SebastianSchuetze** [comment]
Still interesting for me as well!
